### PR TITLE
Erbium: client-side support for CoAP Observe

### PR DESCRIPTION
**Pull request by alessandrelli**
This pull request extends Erbium with client-side support for CoAP Observe.
An example showing how to use the new functionality is also included.

As requested by @mkovatsc, the new code is not compiled by default. To enable it the `COAP_OBSERVE_CLIENT` macro must be defined equal to 1. On a Tmote Sky, this will increase the Text section by ~1100 bytes and the BSS section by ~300 bytes.

Happy Holidays! :)

### CODE REVIEW

File: apps/er-coap/er-coap-observe-client.c [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+  coap_packet_t *pkt;
+  const uint8_t *token;
+  int token_len;
+  coap_observee_t *obs;
+  coap_notification_flag_t flag;
+  uint32_t observe;
+
+  PRINTF("coap_handle_norification()\n");
+  pkt = (coap_packet_t *) notification;
+  token_len = coap_get_token(pkt, &token);
```
Comment (mkovatsc) [inline apps/er-coap/er-coap-observe-client.c]: Use direct packet access in protocol-specific code.

File: apps/er-coap/er-coap-observe-client.c [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+  uint8_t *token;
+  uint8_t token_len;
+  coap_observee_t *obs;
+
+  obs = NULL;
+  coap_init_message(request, COAP_TYPE_CON, COAP_GET, coap_get_mid());
+  coap_set_header_uri_path(request, uri);
+  coap_set_header_observe(request, 0);
+  token_len = coap_generate_token(&token);
+  coap_set_token(request, token, token_len);
```
Comment (mkovatsc) [inline apps/er-coap/er-coap-observe-client.c]: Use direct packet access in protocol-specific code.

**Comment by mkovatsc** [comment]
Sorry for the long wait---thanks @nvt for the reminder!

Looks good, great work @alessandrelli !
The only issue: Remove the coap_*et_token() funtions. These functions are for the protocol-agnostic REST abstraction. The token is protocol-specific...

In the long-run, I will remove this abstraction to achieve a smaller footprint, since there has been no interest for an HTTP binding. Hence, the other setters and getters will probably be removed as well.

**Comment by nvt** [comment]
@mkovatsc Thanks for the comments!

@alessandrelli Please ensure that the code style matches Contiki's style. You can use tools/code-style/uncrustify-fix-style.sh to do so easily.

**Comment by alessandrelli** [comment]
Thanks @mkovatsc and @nvt for the feedback.

During the weekend I will fix the code and update the pull request.

**Comment by alessandrelli** [comment]
Yesterday I have updated the pull request. It should be fine now. If not, just let me know ;)

File: apps/er-coap/er-coap-observe-client.c [hunk 3]
```diff
@@ -0,0 +1,310 @@
+/*
```
Comment (nvt) [inline apps/er-coap/er-coap-observe-client.c]: This file is missing a copyright and license statement.

File: apps/er-coap/er-coap-observe-client.c [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+                         coap_packet_t *notification)
+{
+  coap_packet_t *pkt;
+  const uint8_t *token;
+  int token_len;
+  coap_observee_t *obs;
+  coap_notification_flag_t flag;
+  uint32_t observe;
+
+  PRINTF("coap_handle_norification()\n");
```
Comment (nvt) [inline apps/er-coap/er-coap-observe-client.c]: Typo: "norification"

File: apps/er-coap/er-coap-engine.c [hunk 5]
```diff
@@ -253,6 +253,16 @@ coap_receive(void)
         }
         /* if(ACKed transaction) */
         transaction = NULL;
+
+#if COAP_OBSERVE_CLIENT
+	/* if observe notification */
+        if ((message->type == COAP_TYPE_CON || message->type == COAP_TYPE_NON)
```
Comment (nvt) [inline apps/er-coap/er-coap-engine.c]: There should be no space after "if".

**Comment by nvt** [comment]
@alessandrelli Thanks for the update. I noticed a few minor issues in the source on a second look, and if you can fix these, I think we will be ready to merge this.

**Comment by alessandrelli** [comment]
@nvt, thanks for helping me in having this pull request merged.
I just made the modifications you requested. If anything else is needed, just let me know.

**Comment by nvt** [comment]
:+1:

**Pull request merged**